Migrate TypeScript Configuration to its own package

kmcgrady opened the pull request:
## Describe your changes

Refactor the TypeScript Configuration into a separate package and re-use it in both app and lib.

While working on this, I helped resolved some type issues that soon arose (Not sure what led to them all of a sudden failing, but they seemed valid things to fix).

The build for lib was not completely building the library. I modified the vite config to build the library properly with types and all. Things seemed to look good.

Note: The fact that we use @streamlit/lib/src causes confusion between the node module and the alias. Specifically eslint was failing with regard to private variables. I removed it to make this PR easy. I modify the name in #10213 and re-introduce the private variables. I can merge it first, so it's all clean.

## Testing Plan

- All tests should pass.

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

kmcgrady committed:
Typescript Config

kmcgrady committed:
Do not run typescript checker on tests

mayagbarnes commented on frontend/app/tsconfig.json:
```diff
@@ -1,27 +1,11 @@
+  "root": true,
+  "extends": "@streamlit/typescript-config/react.tsconfig.json",
   "compilerOptions": {
-    "allowJs": true,
```
Are we removing all these compilerOptions because they are on by default or...?

kmcgrady replied:
Yea they pull from tsconfig and compilerOptions are merged. So you can look at react.config (which has the jsx param), and also the base.config (which contains everything else.

kmcgrady committed:
Remove dev.tsconfig.json

kmcgrady committed:
Modify yarn.lock

kmcgrady committed:
Handle typescript errors

sfc-gh-bnisco commented on frontend/lib/src/components/elements/DeckGlJsonChart/useDeckGl.tsx:
```diff
@@ -19,7 +19,6 @@ import { useCallback, useEffect, useMemo, useState } from "react"
 import JSON5 from "json5"
 import { PickingInfo, ViewStateChangeParameters } from "@deck.gl/core"
 import isEqual from "lodash/isEqual"
-import { TooltipContent } from "@deck.gl/core/dist/lib/tooltip"
```
**question:** Is there something wrong with this type import?

kmcgrady replied:
So ESLint was complaining because it was looking for in the exports property of `@deck.gl/core` and not the files property and causing an error. 

I made a solution that [reverts this](https://github.com/streamlit/streamlit/pull/10188/commits/4c9ca484b1196aa9059fa8f991aeb105c76ad68f) and provides the mapping, so hopefully, that's better.

kmcgrady committed:
Address tooltip

sfc-gh-bnisco reviewed (approved):
Nice work here, LGTM!

kmcgrady committed:
Rename @streamlit/lib to ~/lib in streamlit lib. (#10213)

Renamed the `@streamlit/lib/src` search to `~lib` in the lib package.
Using an alias for the package name causes a lot of confusion in ESLint
and TypeScript, and our vite configuration is a little confusing.

I chose the ~lib because that seems natural and shows what's a local
file really easily.

- Tests should pass

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.
